singularity-eos/eos/eos_carnahan_starling.hpp

```diff
@@ -0,0 +1,262 @@
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_EOS_CARNAHAN_STARLING_HPP_
+#define _SINGULARITY_EOS_EOS_EOS_CARNAHAN_STARLING_HPP_
+
+// stdlib
+#include <cmath>
+#include <cstdio>
+#include <string>
+
+// Ports-of-call
+#include <ports-of-call/portability.hpp>
+
+// Base stuff
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/base/robust_utils.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+
+class CarnahanStarling : public EosBase<CarnahanStarling> {
+ public:
+  CarnahanStarling() = default;
+  PORTABLE_INLINE_FUNCTION CarnahanStarling(Real gm1, Real Cv, Real bb, Real qq)
+      : _Cv(Cv), _gm1(gm1), _bb(bb), _qq(qq), _T0(ROOM_TEMPERATURE),
+        _P0(ATMOSPHERIC_PRESSURE), _qp(0.0),
+        _rho0(DensityFromPressureTemperature(_P0, _T0)), _vol0(robust::ratio(1.0, _rho0)),
+        _sie0(Cv * _T0 + qq),
+        _bmod0(_rho0 * Cv * _T0 *
+               (PartialRhoZedFromDensity(_rho0) +
+                ZedFromDensity(_rho0) * ZedFromDensity(_rho0) * gm1)),
+        _dpde0(_rho0 * ZedFromDensity(_rho0) * gm1) {
+    checkParams();
+  }
+  PORTABLE_INLINE_FUNCTION CarnahanStarling(Real gm1, Real Cv, Real bb, Real qq, Real qp,
+                                            Real T0, Real P0)
+      : _Cv(Cv), _gm1(gm1), _bb(bb), _qq(qq), _T0(T0), _P0(P0), _qp(qp),
+        _rho0(DensityFromPressureTemperature(P0, T0)), _vol0(robust::ratio(1.0, _rho0)),
+        _sie0(Cv * T0 + qq),
+        _bmod0(_rho0 * Cv * T0 *
+               (PartialRhoZedFromDensity(_rho0) +
+                ZedFromDensity(_rho0) * ZedFromDensity(_rho0) * gm1)),
+        _dpde0(_rho0 * ZedFromDensity(_rho0) * gm1) {
+    checkParams();
+  }
+  CarnahanStarling GetOnDevice() { return *this; }
+  PORTABLE_INLINE_FUNCTION Real TemperatureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return std::max(robust::SMALL(), (sie - _qq) / _Cv);
+  }
+  PORTABLE_INLINE_FUNCTION Real ZedFromDensity(const Real rho,
+                                               Real *lambda = nullptr) const {
+    const Real eta = _bb * rho;
+    const Real zed = robust::ratio(1.0 + eta + eta * eta - eta * eta * eta,
+                                   (1.0 - eta) * (1.0 - eta) * (1.0 - eta));
+    return zed;
+  }
+  PORTABLE_INLINE_FUNCTION Real PartialRhoZedFromDensity(const Real rho,
+                                                         Real *lambda = nullptr) const {
+    const Real eta = _bb * rho;
+    return robust::ratio(eta * eta * eta * eta - 4.0 * eta * eta * eta + 4.0 * eta * eta +
+                             4.0 * eta + 1.0,
+                         (1.0 - eta) * (1.0 - eta) * (1.0 - eta) * (1.0 - eta));
+  }
+  PORTABLE_INLINE_FUNCTION Real DensityFromPressureTemperature(
```

Comment:
is there a reason you're using newton-raphson here instead of one of the built in root finders?

Reply:
Let me review the built in root finders and see if that would work better. I have a specific root and function shape in mind, which is why I went for the newton-raphson approach.

Reply:
@Yurlungur using RootFinding1D::findRoot now. I left in the newton method with pre processor flags that will default to the root finding utilities if present. Let me know if it is better to simply remove the old newton method.

Reply:
:+1: if `RootFinding1D::findRoot` works, I would just remove the old newton method, unless there's good reason to use it.

Reply:
Removed